shared/utils/lazy.hpp

```diff
@@ -0,0 +1,74 @@
+#include <atomic>
+#include <cstdint>
+#include <functional>
+#include <optional>
+
+/// @brief Thread-safe container which ensures its value is initialized only once.
+/// @tparam T Type of the contained value.
+template <typename T>
+class Once {
+    std::optional<T> value;
+    std::atomic_uintptr_t state_and_queue;
+
+    void initialize(std::function<T()> const& initializer);
+
+  public:
+    /// @brief Constructs an uninitialized `Once<T>`.
+    constexpr Once() : value(), state_and_queue(0) {}
+
+    Once(Once<T> const&) = delete;
+
+    /// @brief Gets the current value, if it is initialized.
+    std::optional<T const&> get() const {
```

Comment:
It's not valid for an optional to hold a reference type. It also doesn't really make sense here, as you literally just return `value` right? `value` isn't an optional wrapping a reference type, so why should the return type for this (and the next one) be that way?